File: ios/hybrids/NitroPlayerSource/HybridNitroPlayerSourceFactory.swift
//
//  HybridNitroPlayerSourceFactory.swift
//  NitroPlay
//
//  Created by Krzysztof Moch on 23/09/2024.
//

import Foundation

class HybridNitroPlayerSourceFactory: HybridNitroPlayerSourceFactorySpec {
  private func normalizeUri(_ uri: String) -> String {
    guard uri.hasPrefix("/") else {
      return uri
    }

    return URL(fileURLWithPath: uri).absoluteString
  }

  private func isHlsManifest(_ uri: String) -> Bool {
    guard var components = URLComponents(string: uri) else {
      return uri.lowercased().hasSuffix(".m3u8")
    }
    components.query = nil
    components.fragment = nil
    return components.string?.lowercased().hasSuffix(".m3u8") == true
  }

  private func normalizedConfig(from config: NativeNitroPlayerConfig) -> NativeNitroPlayerConfig {
    let normalizedUri = normalizeUri(config.uri)
    let shouldUseHlsProxy = config.transport?.mode != .direct
    let route = shouldUseHlsProxy && isHlsManifest(normalizedUri)
      ? HlsProxyRuntime.shared.resolvePlaybackRoute(url: normalizedUri, headers: config.headers)
      : HlsProxyRouteResolution(url: normalizedUri, isProxying: false)

    return NativeNitroPlayerConfig(
      uri: route.url,
      headers: config.headers,
      metadata: config.metadata,
      startup: config.startup,
      buffer: config.buffer,
      retention: config.retention,
      transport: config.transport,
      preview: config.preview
    )
  }

  func fromNitroPlayerConfig(config: NativeNitroPlayerConfig) throws
    -> any HybridNitroPlayerSourceSpec
  {
    let normalizedInput = NativeNitroPlayerConfig(
      uri: normalizeUri(config.uri),
      headers: config.headers,
      metadata: config.metadata,
      startup: config.startup,
      buffer: config.buffer,
      retention: config.retention,
      transport: config.transport,
      preview: config.preview
    )
    let normalized = normalizedConfig(from: normalizedInput)
    return try HybridNitroPlayerSource(
      config: normalized,
      originalConfig: normalizedInput,
      isProxyRouteActive: normalized.uri != normalizedInput.uri
    )
  }

  func fromUri(uri: String) throws -> HybridNitroPlayerSourceSpec {
    let config = NativeNitroPlayerConfig(
      uri: normalizeUri(uri),
      headers: nil,
      metadata: nil,
      startup: .eager,
      buffer: nil,
      retention: nil,
      transport: nil,
      preview: nil
    )
    return try HybridNitroPlayerSource(config: normalizedConfig(from: config))
  }
}

File: android/src/main/java/com/nitroplay/video/hybrids/videoplayersource/HybridNitroPlayerSourceFactory.kt
package com.margelo.nitro.video

import android.content.Context
import android.net.Uri
import androidx.media3.datasource.RawResourceDataSource
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.nitroplay.hls.HlsProxyRuntime
import java.io.File
import java.util.Locale

internal object HybridNitroPlayerUriNormalizer {
  fun isAndroidRawResourceName(input: String): Boolean {
    if (input.isBlank()) {
      return false
    }

    return input.all { character -> character in 'a'..'z' || character in '0'..'9' || character == '_' }
  }

  fun normalize(input: String, applicationContext: Context): String {
    val parsedUri = Uri.parse(input)

    if (parsedUri.scheme != null) {
      return parsedUri.toString()
    }

    if (File(input).isAbsolute || input.startsWith("/")) {
      return Uri.fromFile(File(input)).toString()
    }

    if (!isAndroidRawResourceName(input)) {
      return input
    }

    val resId = applicationContext.resources
      .getIdentifier(input, "raw", applicationContext.packageName)

    if (resId == 0) {
      throw IllegalArgumentException("The video resource '$input' could not be found in res/raw")
    }

    val mediaUri = RawResourceDataSource.buildRawResourceUri(resId)

    return mediaUri.toString()
  }
}

@DoNotStrip
class HybridNitroPlayerSourceFactory: HybridNitroPlayerSourceFactorySpec() {
  private val applicationContext: Context
    get() = NitroModules.applicationContext
      ?: throw IllegalStateException(
        "NitroModules.applicationContext has not been initialized."
      )

  private fun normalizeUri(input: String): String {
    return HybridNitroPlayerUriNormalizer.normalize(input, applicationContext)
  }

  private fun isHlsManifest(uri: String): Boolean {
    val withoutHash = uri.substringBefore('#')
    val withoutQuery = withoutHash.substringBefore('?')
    return withoutQuery.lowercase(Locale.ROOT).endsWith(".m3u8")
  }

  private fun normalizeConfig(config: NativeNitroPlayerConfig): NativeNitroPlayerConfig {
    val normalizedUri = normalizeUri(config.uri)
    val normalizedConfig = config.copy(uri = normalizedUri)
    val shouldUseHlsProxy = normalizedConfig.transport?.mode != NitroSourceTransportMode.DIRECT
    val route = if (shouldUseHlsProxy && isHlsManifest(normalizedUri)) {
      HlsProxyRuntime.resolvePlaybackRoute(normalizedUri, normalizedConfig.headers)
    } else {
      null
    }

    return NativeNitroPlayerConfig(
      uri = route?.url ?: normalizedUri,
      headers = normalizedConfig.headers,
      metadata = normalizedConfig.metadata,
      startup = normalizedConfig.startup,
      buffer = normalizedConfig.buffer,
      retention = normalizedConfig.retention,
      transport = normalizedConfig.transport,
      preview = normalizedConfig.preview
    )
  }

  override fun fromUri(uri: String): HybridNitroPlayerSourceSpec {
    val config = normalizeConfig(NativeNitroPlayerConfig(
      uri = normalizeUri(uri),
      headers = null,
      metadata = null,
      startup = NitroSourceStartup.EAGER,
      buffer = null,
      retention = null,
      transport = null,
      preview = null
    ))

    return HybridNitroPlayerSource(config)
  }

  override fun fromNitroPlayerConfig(config: NativeNitroPlayerConfig): HybridNitroPlayerSourceSpec {
    val normalizedConfig = config.copy(uri = normalizeUri(config.uri))
    val effectiveConfig = normalizeConfig(config)
    return HybridNitroPlayerSource(
      config = effectiveConfig,
      originalConfig = normalizedConfig,
      isProxyRouteActive = effectiveConfig.uri != normalizedConfig.uri
    )
  }

  override val memorySize: Long
    get() = 0
}
